### PR TITLE
More documentation and simplification of BTreeMap's internals

### PR DESCRIPTION
**Pull request by ssomers**
Salvage the documentation and simplification from #67980, without changing the type locked down by debuginfo.

r? @rkruppe

### CODE REVIEW

File: src/liballoc/collections/btree/node.rs [hunk 1]
```diff
@@ -192,17 +195,15 @@ unsafe impl<K: Sync, V: Sync> Sync for Root<K, V> {}
 unsafe impl<K: Send, V: Send> Send for Root<K, V> {}
 
 impl<K, V> Root<K, V> {
+    /// Whether this owned tree is a shared root: the owned `Root` instance wraps
+    /// a bit of shared (not owned) storage.
     pub fn is_shared_root(&self) -> bool {
```
Comment (ssomers) [inline src/liballoc/collections/btree/node.rs]: One extra comment, because I got confused myself about this.

**Comment by JohnTitor** [comment]
I believe you mean r? @hanna-kruppe

File: src/liballoc/collections/btree/node.rs (hunk 1 above)
Comment (ssomers) [reply src/liballoc/collections/btree/node.rs]: Turned into a few comments, that I am happy about.

**Pull request merged**